backend/src/v4/models/modelSetting.js

```diff
@@ -413,58 +420,48 @@ ModelSetting.updateHeliSpeed = async function(account, model, newSpeed) {
 };
 
-ModelSetting.updateMultiplePermissions = async function(account, modelIds, updatedData) {
-	const modelsList = await ModelSetting.findModelSettings(account, {"_id" : {"$in" : modelIds}});
+ModelSetting.updatePermissions = async function(account, model, permissions = []) {
+	const { findByUserName, teamspaceMemberCheck } = require("./user");
 
-	if (!modelsList.length) {
-		throw responseCodes.MODEL_INFO_NOT_FOUND;
+	if (!Array.isArray(permissions)) {
+		throw responseCodes.INVALID_ARGUMENTS;
 	}
 
-	const modelsPromises = modelsList.map((model) => {
-		const newModelPermissions = updatedData.find((modelPermissions) => modelPermissions.model === model._id);
-		return ModelSetting.changePermissions(account, model._id, newModelPermissions.permissions || {});
-	});
-
-	const models = await Promise.all(modelsPromises);
-	const populatedPermissionsPromises = models.map(({permissions}) => {
-		return ModelSetting.populateUsers(account, permissions);
-	});
-	const populatedPermissions = await Promise.all(populatedPermissionsPromises);
-
-	return populatedPermissions.map((permissions, index) => {
-		const {name, federate, _id: model, subModels} =  models[index] || {};
-		return {name, federate, model, permissions, subModels};
-	});
-};
-
-ModelSetting.updatePermissions = async function(account, model, permissions = []) {
 	const setting = await ModelSetting.findModelSettingById(account, model);
-
 	if (!setting) {
 		throw responseCodes.MODEL_NOT_FOUND;
 	}
 
-	permissions.forEach((permissionUpdate) => {
-		if (!setting.permissions) {
-			setting.permissions = [];
-		}
-
-		const userIndex = setting.permissions.findIndex(x => x.user === permissionUpdate.user);
+	permissions = _.uniq(permissions, "user");
+	if (!setting.permissions) {
+		setting.permissions = [];
+	}
 
-		if (-1 !== userIndex) {
-			if ("" !== permissionUpdate.permission) {
-				setting.permissions[userIndex].permission = permissionUpdate.permission;
-			} else {
-				setting.permissions.splice(userIndex, 1);
+	const promises = [];
+	const dbUser = await findByUserName(account);
+	permissions.forEach(permission => {
+		checkPermissionIsValid(permission);
+		checkUserHasPermissionTemplate(dbUser, permission);
+
+		promises.push(teamspaceMemberCheck(permission.user, dbUser.user).then(() => {
+			const index = setting.permissions.findIndex(x => x.user === permission.user);
+			const action = {};
+			if (index !== -1) {
+				if (permission.permission) {
+					setting.permissions[index].permission = permission.permission;
+					action.$set = { permissions: setting.permissions };
+				} else {
+					action.$pull = { permissions: setting.permissions[index] };
+				}
+			} else if (permission.permission) {
+				action.$push = { permissions: permission };
 			}
-		} else if ("" !== permissionUpdate.permission) {
-			setting.permissions.push(permissionUpdate);
-		}
+			db.updateOne(account, MODELS_COLL, { _id: model }, action);
```

Comment:
if we have n permissions, we're doing n updates to the single document.  don't think this is a good idea


Also the fact that you could be overwriting the whole array with $set at L452 means you might end up in a weird state where some $push/pull gets overwritten.

You are probably better off doing a single $set at the end of the loop at this rate.